test/extended/router/metrics.go

```diff
@@ -366,13 +374,34 @@ func findCountersWithLabels(f *dto.MetricFamily, promLabels map[string]string) [
 }
 
 func findGaugesWithLabels(f *dto.MetricFamily, promLabels map[string]string) []float64 {
+	return findGaugesWithLabelsCond(f, promLabels, nil)
+}
+
+func findGaugesWithLabelsCond(f *dto.MetricFamily, promLabels map[string]string, cond func(v float64) bool) []float64 {
 	var result []float64
 	for _, m := range findMetricsWithLabels(f, promLabels) {
-		result = append(result, m.Gauge.GetValue())
+		if cond == nil || cond(m.Gauge.GetValue()) {
+			result = append(result, m.Gauge.GetValue())
+		}
 	}
 	return result
 }
 
+// findNonZeroGaugesWithLabels skips gauges with 0 value to avoid false negatives
+// from the server slots which are DOWN.
+// The Dynamic Configuration Manager (DCM) adds dynamic server slots and
+// uses both these slots and the servers added during HAProxy config rendering.
+// When DCM is enabled, not all server slots are always in the UP state,
+// unlike the behavior without DCM where all slots are UP by default.
+// As a result, we may want to filter out servers where the gauge is set to 0 (== DOWN).
+func findNonZeroGaugesWithLabels(f *dto.MetricFamily, promLabels map[string]string) []float64 {
```

Comment:
Would it be more precise to filter out dynamic servers based on their name? I know we have functions to include labels, so you would need a function to exclude labels. The names of the servers are `_dynamic-pod-#` (2 right now). Could be a regex, or a `contains` type of thing.

Do you think that filtering out zero value gauges here is "weakening" the precision of the test, and opening possibilities for metrics bugs to slip through this E2E test? What happens if haproxy is writing a bunch of bogus metrics for servers, but they have zero value? Would that slip through?

Reply:
Upd: added a `TODO` comment.

Reply:
Sounds like a reasonable future approach. :+1: